File: 1416. Restore The Array.py
# 1416. Restore The Array
class Solution:
    def numberOfArrays(self, s: str, k: int) -> int:
        mod = 10**9 + 7
        n = len(s)
        dp = [0] * (n+1)
        dp[-1] = 1

        for i in range(n-1, -1, -1):
            if s[i] == "0":
                continue
            for j in range(i, n):
                val = int(s[i:j+1])
                if val > k:
                    break
                else:
                    dp[i] += dp[j+1] % mod
        
        return dp[0] % mod

File: 1416. Restore The Array.cpp
// 1416. Restore The Array
#include <string>
#include <vector>
using namespace std;

class Solution {
    public:
        int numberOfArrays(string s, int k) {
            const int mod = 1e9 + 7;
            int n = s.size();
            vector<int> dp(n+1, 0);
            dp[n] = 1;
    
            for (int i = n-1; i > -1; i--) {
                if (s[i] == '0') continue;
                long long num = 0;
                for (int j = i; j < n; j++) {
                    num = num * 10 + (s[j] - '0');
                    if (num > k) break;
                    dp[i] = (dp[i] + dp[j+1]) % mod;
                }
            }
    
            return dp[0] % mod;
    
        }
    };
